core/src/main/scala/org/ensime/core/AnalyzerManager.scala

```diff
@@ -0,0 +1,381 @@
+package org.ensime.core
+
+import java.nio.file.Path
+
+import akka.actor.{ Actor, ActorLogging, ActorRef, Props, Stash }
+import akka.event.LoggingReceive.withLabel
+import org.ensime.api._
+import org.ensime.config._
+import org.ensime.indexer.SearchService
+import org.ensime.util.FileUtils.toSourceFileInfo
+import org.ensime.util.ensimefile._
+import org.ensime.util.file.File
+import org.ensime.vfs.`package`.EnsimeVFS
+
+import scala.collection.{ breakOut, mutable }
+
+trait ModuleFinder {
+  implicit val config: EnsimeConfig
+
+  private val module_srcs = for {
+    project <- config.projects
+    source <- project.sources
+  } yield (project, source.toPath)
+
+  def getModule(path: Path): Option[EnsimeProject] = {
+    val module_tuple = module_srcs.find {
+      case (name, source) => path.toString.startsWith(source.toString)
+    }
+    module_tuple match {
+      case Some((module, _)) =>
+        Some(module)
+      case None =>
+        None
+    }
+  }
+
+  def getModule(fileInfo: SourceFileInfo): Option[EnsimeProject] = fileInfo.file match {
+    case RawFile(f) => getModule(f)
+    case ArchiveFile(f, _) => getModule(f)
+  }
+  def getModule(file: File): Option[EnsimeProject] = getModule(file.toPath)
+  def getModule(file: Either[File, SourceFileInfo]): Option[EnsimeProject] = file match {
+    case Left(f) => getModule(f)
+    case Right(sourceFileInfo) => getModule(sourceFileInfo)
+  }
+}
+
+sealed trait FileStatus {
+  def append(status: FileStatus): FileStatus
+}
+final case object NotLoaded extends FileStatus {
+  def append(status: FileStatus) = status match {
+    case Loaded => Loaded
+    case _ => NotLoaded
+  }
+}
+final case object Loaded extends FileStatus {
+  def append(status: FileStatus) = status match {
+    case Removed => Removed
+    case Unloaded => Unloaded
+    case _ => Loaded
+  }
+}
+final case object Removed extends FileStatus {
+  def append(status: FileStatus) = status match {
+    case Loaded => Loaded
+    case _ => Removed
+  }
+}
+final case object Unloaded extends FileStatus { // different from Removed
+  def append(status: FileStatus) = status match {
+    case Loaded => Loaded
+    case _ => Unloaded
+  }
+}
+
+final case class LoadedFilesData(statusOfFile: Map[SourceFileInfo, FileStatus], default: FileStatus)
+
+class AnalyzerManager(
+    broadcaster: ActorRef,
+    indexer: ActorRef,
+    search: SearchService,
+    implicit val config: EnsimeConfig,
+    implicit val vfs: EnsimeVFS
+) extends Actor with ActorLogging with ModuleFinder with Stash {
+
+  // maps the active modules to their analyzers
+  private var analyzers: mutable.Map[EnsimeProject, ActorRef] = mutable.Map()
```

Comment:
or a `var` of an immutable `Map` (which I find better because then it's explicit). But doing both is very unusual.